### ISSUE TITLE
Highlight all matching occurrences of text in selection

### ISSUE THREAD

**Issue by hoanganh2303** [zed-industries/zed#12635]
### Check for existing issues

- [x] Completed

### Describe the feature

highlight same values in zed editor

**Comment by JosephTLyons** [comment]
Hey @hoanganh2303 - I'm not entirely sure I understand what you are looking for - does this feature exist in another editor? If so, could you share a screenshot here?

**Comment by jaydenseric** [comment]
Related discussion: https://github.com/zed-industries/zed/discussions/9238 .

**Comment by elawad** [comment]
Hi I’m new to Zed and also came looking for this solution.

In Atom you could double-click to highlight a string/variable. Atom will highlight all occurrences in the file. 

Zed only highlights direct usage of a variable.
Not in the entire file, and doesn’t work on strings etc.

**Comment by lanternlogic** [comment]
Yeah this is the sole reason I've not switched entirely to Zed for small scripts and viewing the odd files that VSCode would be too cumbersome for.
Its incredibly useful to be able to see other occurences of names/variables/tokens just by highlighting one of them.

**Comment by elawad** [comment]
Current workaround, though not perfect:

1. Click variable
2. Cmd + F (find)

**Comment by JosephTLyons** [comment]
Ah I see - we have this implemented for when you select symbols in Zed, but not for arbitrary text.

**Comment by lanternlogic** [comment]
On a similar note, even for files that are supported I do not think the highlighted occurences of selected word are obvious enough. Would this be more a theme-based thing? or something separate to do with this highlighting feature. _Really_ looking forward to this.

**Comment by AndrStnz** [comment]
This is a very important comfort feature, please implement this :)

**Comment by lanternlogic** [comment]
beating a dead horse here, but is there any update!

_Man_ it would be nice to have here
<img width="650" alt="Screenshot 2024-07-23 at 12 36 53 pm" src="https://github.com/user-attachments/assets/b02187ee-fb14-47bb-9307-c2c62a57e027">

**Comment by lanternlogic** [comment]
beating a dead horse, but is there any update 😄 

_Man_ it would be nice to have here
<img width="650" alt="Screenshot 2024-07-23 at 12 36 53 pm" src="https://github.com/user-attachments/assets/b6e90009-c06f-43ef-beef-dc916ed927dd">

**Comment by g0t4** [comment]
Here is the vscode impl: https://github.com/microsoft/vscode/blob/main/src/vs/editor/contrib/multicursor/browser/multicursor.ts#L847-L848

IIAC this is going to be language specific? Or can it be added across any file type? I'd be happy to help with a bit of guidance on a proper zed implementation.

**Comment by btrazzini** [comment]
I really need it for any text, not just for symbols.

![Print 20240924-08 00 08](https://github.com/user-attachments/assets/214d1202-c5fc-42d7-87af-673fa789e98a)

**Comment by Amplificator** [comment]
I'd love to have this feature as well for any selected text.

**Comment by shinebayar-g** [comment]
I never realized small things like this are what makes VSCode so good and hard to replace with Zed. _Details matter_.

**Comment by Seltonu** [comment]
This seems to be partially working now, specifically for PHP code only. Double clicking i.e. a variable name will highlight other occurrences. Didn't see anything mentioned in the release notes about it

**Comment by danielweck** [comment]
I agree, this is one of the few reasons I still occasionally use Visual Studio Code instead of Zed :)
My editing workflow regularly includes selecting arbitrary text in order to instantly (i.e. zero UX friction) visualise verbatim matches inside the current document's scrolling viewport. The search feature cannot be used to emulate this because I need to expand/contract the text selection by character or other units (in my case, using the keyboard) in order to see matching highlights instantly in the vicinity.
I realise that this behaviour might actually be annoying to some users, but I'd be happy with an opt-in :)
Thank you!

**Comment by nickmachairas** [comment]
Adding my voice here. Subtle but very handy. I hope it's added soon.

**Comment by Kociamber** [comment]
It's one of those few features that makes me stick with VS code. Guys, please add it 🙏🏻

**Comment by 497e0bdf29873** [comment]
This is essential for doing e.g. *local* symbol renames by selecting the first occurrence of the symbol (double-click), and then using `cmd+D` to select the rest. You want to see easily see how many times to press `cmd+D`.

**Comment by g0t4** [comment]
Thanks for adding this!

Question, it seems to only work for selections that don't span multiple lines?

https://github.com/user-attachments/assets/a0953217-dbec-476c-b199-708c1ef12592

Also, minor consideration... when using mouse to click and drag to select a range... it doesn't show matches until you release the click. After that it shows changes to the selection as you make it 

https://github.com/user-attachments/assets/ddb31b78-ecdf-47b6-afc3-4bd350d21385

Last, a small recommendation... [visimatch.nvim](https://github.com/wurli/visimatch.nvim) added support for whitespace mismatches, IOTW the whitespace can differ so long as the selected text otherwise remains the same. Especially nice for indentation differences, to still spot overlap.

**Comment by danielweck** [comment]
good point about whitespace and line break matching ... and thank you Zed devs for implementing this, much appreciated :)

**Comment by 0xtimsb** [comment]
@g0t4 to answer your questions:

1. Yes, it currently only works for a single line. Ideally, we should handle multiple lines in case of wrapping.
2. We have a default debounce time of `50 ms` for this. If you'd like it to happen in real-time, setting it to `0` will do the trick.
3. Hmm, gotta check out [visimatch.nvim](https://github.com/wurli/visimatch.nvim).

I'm not actively addressing the above mentioned points right now, but I'll eventually get to them as time allows or if an issue comes up.

**Comment by Seltonu** [comment]
In my testing I see two limitations with the current implementation

- Highlight matching is dependent on syntax, NOT matching any string of text. Personally I consider this very important, I'm often searching for random strings


- Matches do not occur on unsaved changes

![Image](https://github.com/user-attachments/assets/717f93bc-ebbe-410c-858c-71d9a7e4bbb8)

**Comment by 0xtimsb** [comment]
@Seltonu Are you on the latest preview of Zed? This feature isn’t in stable yet, and I can’t reproduce it on preview. If you’re sure, please open a separate issue to track it.

**Comment by Seltonu** [comment]
@0xtimsb Oh! Sorry for the confusion, I'm running the latest stable (v0.174.6) on Linux. Bit odd that it does highlighting then but I just tested on latest preview and can confirm these issues aren't present. I apologize for the trouble haha, I thought it made it out already. Looking forward to this hitting stable!

**Comment by xgfone** [comment]
> - Highlight matching is dependent on syntax, NOT matching any string of text. Personally I consider this very important, I'm often searching for random strings. It also appears to not work for matching HTML/CSS classes because of this.
> - Matches do not occur on unsaved changes

I had tested it on the `main` branch, both of them are ok.
```
Zed: v0.176.0 (Zed Dev ee280b0d058af3fc5705133d6531f8860c73498a) 
```

**Comment by lanternlogic** [comment]
Definitely a dumb question and am new to this OSS scene, but where would I download the preview release which is linked to this issue?

**Comment by SomeoneToIgnore** [comment]
See https://github.com/zed-industries/zed?tab=readme-ov-file#installation for installation, also there's https://github.com/zed-industries/zed/releases with all releases and related changes enumerated.

**Comment by Seltonu** [comment]
One issue I found is that highlighting matching does not match by case, didn't see it mentioned in this thread yet. See example with lowercase "Html" getting matched where I expect it wouldn't. I was getting an undefined error in browser console for the variable that confused me for a while because the matching was a bit misleading. Is there a relevant setting for this, or just unsupported currently?

![Image](https://github.com/user-attachments/assets/e2032fe6-4132-4184-a889-be1256928b50)

I feel that this scenario perfectly captures a common use case for me - checking for typos in variables/strings/paths with quick highlight matching, and so matching by case is important

**Comment by AndrStnz** [comment]
![Image](https://github.com/user-attachments/assets/2853e87a-8ff2-489c-82d0-a6eecc760b67)
Oh this is really awkward. Yes, matching should be case sensitive. I am not in Windows Land here....

**Comment by teohhanhui** [comment]
> Oh this is really awkward. Yes, matching should be case sensitive. I am not in Windows Land here....

It should just follow the current / last used Find case sensitivity, i.e. case insensitive by default. That's also how it works in vscode.

**Comment by 497e0bdf29873** [comment]
I just noticed that this doesn't quite work as expected: If I select a something, it first highlights all the matches, as expected. However, once I use `Cmd+D` to select also the second match, it no longer highlights the remaining matches. So it still isn't easy to use `Cmd+D` to select all local matches.
